### PR TITLE
Settings app

### PR DESCRIPTION
**Pull request by marekventur**
A simple settings app that allows users to
* change their name (as displayed on the homescreen)
* change wifi details
* change badge store settings
* pick a startup app

I've wrapped some of the common functionality into a lib - I'm planning to roll it out to the badge store as well, since it has a similar navigation structure

Fixes https://github.com/emfcamp/Mk4-Apps/issues/5

### CODE REVIEW

File: settings/badge_store_settings.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+from stack_nav import *
+from database import *
```
Comment (jburgess777) [inline settings/badge_store_settings.py]: does this need license or dependencies set?

File: lib/stack_nav.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            stack = [(init_fn, init_state)]
+        elif result == NEXT_EXIT:
+            break
+        else:
+            stack.pop()
+
+# A simple menu. Format of items is [{"foo":fn}, ...]
+def selection(items, title="Settings", none_text="Back"):
+    items = [{"title": t, "function": fn} for (t, fn) in items.items()]
+    selection = dialogs.prompt_option(items, none_text=none_text, title="Settings")
```
Comment (jburgess777) [inline lib/stack_nav.py]: maybe `title=title` was intended?

File: settings/main.py [hunk 3]
```diff
@@ -6,71 +6,41 @@
...
-
-    print("bye")
-
-# Entry point
+    return selection({
+        "Homescreeen Name": change_database_string("Set your name", "name"),
+        "Wifi": settings_wifi,
+        "Startup app": settings_startup_app,
+        "Badge Store": settings_badge_store
+    }, title="Settings", none_text="Exit")
```
Comment (jburgess777) [inline settings/main.py]: Is `title="Settings"` here redundant?

File: lib/dialogs.py [hunk 4]
```diff
@@ -113,7 +113,7 @@ def prompt_text(description, init_text = "", true_text="OK", false_text="Back",
         edit.destroy();
     return
 
-def prompt_option(options, index=0, text = "Please select one of the following:", title=None, select_text="OK", none_text=None):
+def prompt_option(options, index=0, text = None, title=None, select_text="OK", none_text=None):
```
Comment (thibault-ml) [inline lib/dialogs.py]: Should be `text=None` (spacing)

File: lib/stack_nav.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+___dependencies___ = ["dialogs", "database"]
+___license___ = "MIT"
+
+import dialogs, database
+
+NEXT_EXIT = "exit" # Leave navigation
+NEXT_TOP = "init" # Go to top of stack
+def nav(init_fn, init_state={}):
+    stack = [(init_fn, init_state)]
+    while len(stack):
```
Comment (thibault-ml) [inline lib/stack_nav.py]: empty arrays evaluate as `False`, so `len` is not required. The following should be enough:

```python
while stack:
    # ...
```

File: settings/badge_store_settings.py (hunk 1 above)
Comment (marekventur) [reply settings/badge_store_settings.py]: no, only main.py needs dependencies for apps (I should document that)

File: lib/stack_nav.py (hunk 2 above)
Comment (marekventur) [reply lib/stack_nav.py]: good catch!

File: settings/main.py (hunk 3 above)
Comment (marekventur) [reply settings/main.py]: 👍

**Comment by marekventur** [comment]
@thibault-ml, @jburgess777 - who are you on IRC? I'm trying to find some people we can ask for help during the event with reviewing apps and PRs in general. Would you be interested?

**Comment by thibault-ml** [comment]
@marekventur I'm not much of an IRC person, but happy to help where needed :-)

**Comment by jburgess777** [comment]
@marekventur I am `jburgess` on IRC. I will be at the event too and can help.

**Comment by marekventur** [comment]
Thanks! Just drop by the badge tent when you're around!

File: .development/pyboard_util.py [hunk 6]
```diff
@@ -187,4 +187,32 @@ def end_copy_via_repl(args):
...
 def clean_via_repl(args):
-    raise Exception("not implemented yet")
+    init_copy_via_repl(args)
+    print("Cleaning:", end=" ", flush=True)
+    try:
+        execbuffer(get_pyb(args), "clean()")
+    except PyboardError as er:
+        print("FAIL")
+        print(er)
+        pyb.close()
```
Comment (jburgess777) [inline .development/pyboard_util.py]: is `pyb` defined here? The equivalent `hard_reset()` function reads it before the try/except block.

Comment (marekventur) [reply .development/pyboard_util.py]: sorry, this ended up in the wrong branch - will fix :(

Comment (marekventur) [reply .development/pyboard_util.py]: pushed all commits to the correct branch now, sorry for the confusion.

**Pull request merged**